File: src/test/java/be/angelcorp/libs/celest/kepler/TestKeplerEquations.java
/**
 * Copyright (C) 2011 simon <aodtorusan@gmail.com>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *        http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.libs.celest.kepler;

import junit.framework.TestCase;

import org.apache.commons.math.linear.ArrayRealVector;

import be.angelcorp.libs.celest.constants.EarthConstants;
import be.angelcorp.libs.celest.stateVector.CartesianElements;
import be.angelcorp.libs.celest.stateVector.ICartesianElements;
import be.angelcorp.libs.celest.stateVector.IKeplerElements;
import be.angelcorp.libs.celest.stateVector.KeplerElements;
import be.angelcorp.libs.celest.unit.Tests;
import be.angelcorp.libs.math.MathUtils2;
import be.angelcorp.libs.math.linear.Vector3D;

public class TestKeplerEquations extends TestCase {

	/************************************
	 * Non-Static function tests
	 ************************************/
	// double arealVel()
	// double arguementOfLatitude()
	// double flightPathAngle()
	// CartesianElements kepler2cartesian()
	// double meanMotion()
	// double period()
	// double radius(double p, double e, double nu)
	// double totEnergyPerMass()
	// double trueAnomalyFromAnomaly()
	// double trueAnomalyFromMean()
	// double trueAnomalyFromMean(double M)
	// double trueLongitude()
	// double trueLongitudeOfPeriapse()
	// double trueLongitudeOfPeriapseApproximate()
	// double visViva(double r)

	public void test_Anomaly() {
		// Delegates calls, several samples:
		// Test based on validated samples in ReferenceKeplerAngles class
		for (ArrayRealVector sample : ReferenceKeplerAngles.all) {
			KeplerElements k = ReferenceKeplerAngles.createKeplerElements(sample);
			double anomaly = k.getOrbitEqn().anomaly();
			ReferenceKeplerAngles.checkEqualAnomaly(sample, anomaly);
		}
	}

	public void test_MeanAnomaly() {
		// Delegates calls, several samples:
		// Test based on validated samples in ReferenceKeplerAngles class
		for (ArrayRealVector sample : ReferenceKeplerAngles.all) {
			KeplerElements k = ReferenceKeplerAngles.createKeplerElements(sample);
			double M = k.getOrbitEqn().meanAnomaly();
			ReferenceKeplerAngles.checkEqualMeanAnomaly(sample, M);
		}
	}

	/************************************
	 * Static function tests
	 ************************************/
	public void testStaticArguementOfLatitude() {
		// public static double arguementOfLatitude(double w, double nu)
		// public static double arguementOfLatitude(Vector3D nodalVector, Vector3D radius)
		// TODO: Add tests
	}

	public void testStaticCartesian2kepler() {
		// cartesian2kepler(CartesianElements, CelestialBody)
		// cartesian2kepler(CartesianElements, double)
		ICartesianElements c = new CartesianElements(
				new Vector3D(10157768.1264, -6475997.0091, 2421205.9518),
				new Vector3D(1099.2953996, 3455.105924, 4355.0978095));
		IKeplerElements k = KeplerEquations.cartesian2kepler(c, EarthConstants.mu);
		double[] expected = new double[] {
				1.216495E7, 0.01404, 0.919398, 2.656017, 5.561776, 3.880560 };
		double[] tol = new ArrayRealVector(expected).mapMultiply(1E-3).toArray();
		Tests.assertEquals(expected, k.toVector().getData(), tol);
		k = KeplerEquations.cartesian2kepler(c, EarthConstants.bodyCenter);
		Tests.assertEquals(expected, k.toVector().getData(), tol);
	}

	public void testStaticCartesian2kepler2D() {
		// cartesian2kepler2D(CartesianElements, CelestialBody)
		ICartesianElements c = new CartesianElements(
				new Vector3D(10157768.1264, -6475997.0091, 2421205.9518),
				new Vector3D(1099.2953996, 3455.105924, 4355.0978095));
		IKeplerElements k = KeplerEquations.cartesian2kepler2D(c, EarthConstants.bodyCenter);
		assertEquals(1.216495E7, k.getSemiMajorAxis(), 1E4);
		assertEquals(0.01404, k.getEccentricity(), 1E-5);
		assertEquals(3.88056, k.getTrueAnomaly(), 1E-2);
	}

	public void testStaticEccentricity() {
		// eccentricity(double, double)
		assertEquals(0, KeplerEquations.eccentricity(1, 1), 1E-3);
		// Earth to mars orbit perihelion is the Earth's orbital radius, aphelion is the radius of Mars
		double Rp = 1.495978E11;
		double Ra = 2.27987047E11;
		assertEquals(0.207606972, KeplerEquations.eccentricity(Rp, Ra), 1E-9);
	}

	public void testStaticFix2dOrbit() {
		// fix2dOrbit(KeplerElements)
		IKeplerElements k = new KeplerElements(0, 0, 0, Double.NaN, Double.NaN, 0);
		KeplerEquations.fix2dOrbit(k);
		assertTrue(!Double.isNaN(k.getRaan()));
		assertTrue(!Double.isNaN(k.getArgumentPeriapsis()));
	}

	public void testStaticFlightPathAngle() {
		// flightPathAngle(double, double)
		// http://www.wolframalpha.com/input/?i=mean+anomaly+5+radian+eccentricity+0.6
		double gamma = MathUtils2.mod(KeplerEquations.flightPathAngle(0.6, -2.427), 2 * Math.PI);
		assertEquals(5.6597, gamma, 1E-4);
	}

	// public static Vector3D getH(Vector3D R, Vector3D V)
	// public static Matrix3D gravityGradient(Vector3D R, double mu)

	public void testStaticKepler2Cartesian() {
		// public static CartesianElements kepler2cartesian(double a, double ecc, double inc, double
		// Omega, double w, double nu, double mu)
		ICartesianElements c = KeplerEquations.kepler2cartesian(
				1.216495E7, 0.01404, 0.919398, 2.656017, 5.561776, 3.880560, EarthConstants.mu);
		// Values from keplerCOE from Matlab Orbital_Library by Richard Rieber
		ICartesianElements c_true = new CartesianElements(
				new Vector3D(1.092882447232868e+007, -5.619415989750504e+006, -1.715953308630781e+005),
				new Vector3D(1.466941526515634e+003, +3.108913288555892e+003, -4.504368922790057e+003));

		assertTrue(c_true.equals(c, 1e-12));
	}

	// Vector3D localGravity(Vector3D r, double mu)
	// double meanMotion(double, double)
	// double trueLongitude(Vector3D radius)
	// double trueLongitudeApproximation(double raan, double w, double nu)
	// double trueLongitudeOfPeriapse(Vector3D eccentricityVector)
	// double trueLongitudeOfPeriapseApproximate(double RAAN, double w)
	// double visViva(double, double, double)

}

File: celest/src/test/java/be/angelcorp/celest/kepler/TestKeplerEquations.scala
/**
 * Copyright (C) 2009-2012 simon <simon@angelcorp.be>
 *
 * Licensed under the Non-Profit Open Software License version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.opensource.org/licenses/NOSL3.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package be.angelcorp.celest.kepler

import be.angelcorp.celest.body.Satellite
import be.angelcorp.celest.constants.Constants._
import be.angelcorp.celest.frameGraph.frames.BodyCenteredSystem
import be.angelcorp.celest.kepler
import be.angelcorp.celest.math.geometry.Vec3
import be.angelcorp.celest.math._
import be.angelcorp.celest.state.PosVel
import be.angelcorp.celest.unit.CelestTest
import be.angelcorp.celest.universe.DefaultUniverse
import org.scalatest.FlatSpec
import org.scalatest.matchers.ShouldMatchers

import scala.math._


class TestKeplerEquations extends FlatSpec with ShouldMatchers with CelestTest {

  implicit val universe = new DefaultUniverse

  /** **********************************
    * Non-Static function tests
    * ***********************************/

  "KeplerEquations" should "evaluate the elliptical/parabolic/hyperbolic anomaly correctly from e/M" in {
    // Delegates calls, several samples:
    // Test based on validated samples in ReferenceKeplerAngles class
    ReferenceKeplerAngles.all.foreach({
      sample =>
        val k = ReferenceKeplerAngles.createKeplerElements(sample)
        val anomaly = k.anomaly

        expect(true, "Anomaly not correct for entry %s, computed %f".format(sample, anomaly))(
          ReferenceKeplerAngles.checkEqualAnomaly(sample, anomaly)
        )
    })
  }

  it should "evaluate the true anomay correctly from e/M" in {
    // Delegates calls, several samples:
    // Test based on validated samples in ReferenceKeplerAngles class
    ReferenceKeplerAngles.all.foreach({
      sample =>
        val k = ReferenceKeplerAngles.createKeplerElements(sample)
        val anomaly = k.trueAnomaly

        expect(true, "Mean anomaly not correct for entry %s, computed %f".format(sample, anomaly))(
          ReferenceKeplerAngles.checkEqualTrueAnomaly(sample, anomaly)
        )
    })
  }

  /** **********************************
    * Static function tests
    * ***********************************/

  //it should "compute the correct arguement of latitude" in {
  // public static double arguementOfLatitude(double w, double nu)
  // public static double arguementOfLatitude(Vec3 nodalVector, Vec3 radius)
  // TODO: Add tests
  //}

  it should "compute the correct cartesian to kepler conversion" in {
    val frame = new BodyCenteredSystem {
      val centerBody = new Satellite(5.9736E24, null)
    }
    val c = new PosVel(
      Vec3(10157768.1264, -6475997.0091, 2421205.9518),
      Vec3(1099.2953996, 3455.105924, 4355.0978095),
      frame
    )

    val k = kepler.cartesian2kepler(c, frame.centerBody.μ)
    val expected = Array[Double](1.216495E7, 0.01404, 0.919398, 2.656017, 5.561776, 3.880560)
    val tol = expected.map(_ * 1E-3)

    expected(0) should be (k._1 +- tol(0))
    expected(1) should be (k._2 +- tol(1))
    expected(2) should be (k._3 +- tol(2))
    expected(3) should be (k._4 +- tol(3))
    expected(4) should be (k._5 +- tol(4))
    expected(5) should be (k._6 +- tol(5))
  }

  it should "compute the correct cartesian to kepler (2d) conversion" in {
    val frame = new BodyCenteredSystem {
      val centerBody = new Satellite(5.9736E24, null)
    }
    val c = new PosVel(
      Vec3(10157768.1264, -6475997.0091, 2421205.9518),
      Vec3(1099.2953996, 3455.105924, 4355.0978095),
      frame
    )

    val k = kepler.cartesian2kepler2D(c, frame.centerBody.μ)

    k._1 should be(1.216495E7 +- 1E4)
    k._2 should be(0.01404 +- 1E-5)
    k._3 should be(3.88056 +- 1E-2)
  }

  it should "compute the correct eccentricity from Ra and Rp" in {
    // eccentricity(double, double)
    kepler.eccentricity(1, 1) should be(0.0 +- 1E-16)

    // Earth to mars orbit perihelion is the Earth's orbital radius, aphelion is the radius of Mars
    val Rp = 1.495978E11
    val Ra = 2.27987047E11
    kepler.eccentricity(Rp, Ra) should be(0.207606972 +- 1E-9)
  }

  it should "compute the correct flight path angle" in {
    // http://www.wolframalpha.com/input/?i=mean+anomaly+5+radian+eccentricity+0.6
    val gamma = mod(kepler.flightPathAngle(0.6, -2.427), 2 * Pi)
    gamma should be(5.6597 +- 1E-4)
  }

  it should "compute the kepler elements for cartesian elements" in {
    // Values from keplerCOE from Matlab Orbital_Library by Richard Rieber
    val c = kepler.kepler2cartesian(1.216495E7, 0.01404, 0.919398, 2.656017, 5.561776, 3.880560, mass2mu(5.9736E24))
    val c_true = new PosVel(
      Vec3(1.092882447232868e+007, -5.619415989750504e+006, -1.715953308630781e+005),
      Vec3(1.466941526515634e+003, +3.108913288555892e+003, -4.504368922790057e+003),
      null
    )

    assertEquals( new PosVel(c._1, c._2, null ), c_true, 1E-5, 1E-9)
  }

}
